### PR TITLE
Music search query history

### PR DESCRIPTION
**Pull request by Ben2356**
My long overdue PR for this that I've been working on off and on 😅! This adds the saving a simple history of the queries that the user has previously searched. In order to have a query be saved the user must either 1) click Play from Search, 2) click View Full Results or the ellipsis entry, 3) click on a previous search history query, 4) click on a search result item, 5) click on the OK button to bring up the full page search results . The search history query saves the last 8 searches, the most recent of the search queries being displayed at the bottom of the list. 

Additionally with this feature the search logic has been refactored/updated to help prevent race conditions when displaying search results in the suggestion list and help smoothen out the experience.

I wasn't able to get any videos / pictures of this one but didn't want to hold this up any longer, sorry about that.

### CODE REVIEW

**Comment by hufman** [comment]
Please merge the latest main branch, which should include a fix for running the CI tests in a Pull Request context!

File: app/src/test/java/me/hufman/androidautoidrive/music/MusicAppTest.kt [hunk 1]
```diff
@@ -2066,32 +2070,123 @@ class MusicAppTest {
...
+		verify(musicController).playSong(searchResultItem)
+
+		// search history should be updated
+		val searchQueryHistory = gson.fromJson(musicController.appSettings[AppSettings.KEYS.MUSIC_SEARCH_QUERY_HISTORY], Array<String>::class.java)
+		assertEquals(searchQueryHistory.size, 1)
+		assertEquals(searchQueryHistory[0], searchQuery)
+	}
+
+	@Test
+	fun testSearch_MultipleSearchJobs() {
```
Comment (hufman) [inline app/src/test/java/me/hufman/androidautoidrive/music/MusicAppTest.kt:2107]: Thank you for adding tests for this tricky piece of code!

**Review by hufman — CHANGES_REQUESTED** [review]

**Comment by Ben2356** [comment]
> Please merge the latest main branch, which should include a fix for running the CI tests in a Pull Request context!

Merged the latest!

File: app/src/test/java/me/hufman/androidautoidrive/music/MusicAppTest.kt [hunk 2]
```diff
@@ -170,6 +171,7 @@ class MusicAppTest {
 		on { loadDesiredApp() } doReturn ""
 		on { musicSessions } doReturn mock<MusicSessions>()
 		on { isConnected() } doReturn true
+		on { appSettings } doReturn MockAppSettingsReceiver(mock())
```
[the remark this answers is not in the document]
Comment (Ben2356) [reply app/src/test/java/me/hufman/androidautoidrive/music/MusicAppTest.kt]: Ah you're right, not sure why I didn't update the type of the `MusicController.appSettings` and instead went this route. Thanks for pointing that out

File: app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/SearchInputView.kt [hunk 3]
```diff
@@ -123,12 +188,33 @@ class SearchInputView(val state: RHMIState,
...
+			 * added the oldest search query will be dropped from the collection.
+			 */
+			private fun updateSearchQueryHistory(input: String) {
+				if (searchQueryHistory.contains(input)) {
+					searchQueryHistory.remove(input)
+				}
+				if (searchQueryHistory.size == SEARCH_HISTORY_QUERY_MAX_COUNT) {
+					searchQueryHistory.remove(searchQueryHistory.first())
+				}
+				searchQueryHistory.add(input)
```
[the remark this answers is not in the document]
Comment (Ben2356) [reply app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/SearchInputView.kt]: Yeah, I was originally was using the top of the list as the method docs state but changed this when I was experimenting with the suggest model and never updated it back. Updated to move latest queries to the top of the list

**Review by hufman — APPROVED** [review]

**Pull request merged**